src/twisted/internet/defer.py

```diff
@@ -160,6 +167,87 @@ def maybeDeferred(f, *args, **kw):
 
 
 
+def extractCurrentResult(deferred, raises=True, passthrough=False):
+    """
+    Extract the current result from a L{Deferred} that has already fired all of
+    its callbacks.
+
+    Take a deferred that has already fired all of its callbacks and extract and
+    return the result of the L{Deferred}. This process of extracting the result
+    will, by default, consume the result, error or not unless the
+    L{passthrough} parameter is set to L{True}. If the L{Deferred} is in an
+    error then the process of extracting the result will, by default, re-raise
+    the original exception unless the L{raises} parameter is set to L{False}.
+
+    The primary purpose of this function is to allow code to be written that is
+    internally asynchronous but which can be consumed without an event loop in
+    an synchronous fashion by using L{extractCurrentResult} at the boundaries.
+    This enables libraries to be written that can function both with Twisted
+    and with synchronous code instead of requiring an asynchronous and a
+    synchronous implementation for all functionality.
+
+    @attention: This function CANNOT take a Deferred with pending asynchronous
+        work to be done and make it synchronous. It can ONLY extract results
+        from a L{Deferred} which has already fired its callbacks.
+
+    @type deferred: Deferred
+    @param deferred: The Deferred from which results will be extracted..
+
+    @type raises: bool
+    @param raises: Whether or not to raise the original exception (L{True}) or
+        to return the failure object (L{False}).
+
+    @type passthrough: bool
+    @param passthrough: Whether or not to pass through the original value
+        (L{True}) or to consume any results, error or otherwise (L{False}).
```

Comment:
I'd prefer different functions for the behavior granted by these different flags.  For example, `extractCurrentResult` seems like it should always have the `passthrough=False` behavior.  That's a reasonable interpretation of "extract".  The `passthrough=True` behavior could be exposed separately as (for example) `observeCurrentResult`.  "Observe" is reasonably interpreted as not modifying the target object (except to quantum physicists of course :wink:).


Providing _four_ different APIs for all combos of the two flags is probably ridiculous and unhelpful and I don't have any suggestions for the names of things along the `raises` dimension anyway, so (unless you have some ideas) just fixing this for `passthrough` would be okay with me.

While I have no doubt this would produce a better API, this is clearly a bikeshed.  I leave the decision to you.